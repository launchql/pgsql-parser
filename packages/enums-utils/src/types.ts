import { format } from 'util';

const fail = (type, node) => {
  throw new Error(format('Unhandled %s node: %s', type, JSON.stringify(node)));
};

export const TYPE_NAMES = {
  OBJECT_ACCESS_METHOD: 'ACCESS METHOD',
  OBJECT_AGGREGATE: 'AGGREGATE',
  OBJECT_ATTRIBUTE: 'ATTRIBUTE',
  OBJECT_CAST: 'CAST',
  OBJECT_COLLATION: 'COLLATION',
  OBJECT_COLUMN: 'COLUMN',
  OBJECT_CONVERSION: 'CONVERSION',
  OBJECT_DATABASE: 'DATABASE',
  OBJECT_DOMAIN: 'DOMAIN',
  OBJECT_DOMCONSTRAINT: 'CONSTRAINT',
  OBJECT_EVENT_TRIGGER: 'EVENT TRIGGER',
  OBJECT_EXTENSION: 'EXTENSION',
  OBJECT_FDW: 'FOREIGN DATA WRAPPER',
  OBJECT_FOREIGN_SERVER: 'SERVER',
  OBJECT_FOREIGN_TABLE: 'FOREIGN TABLE',
  OBJECT_FUNCTION: 'FUNCTION',
  OBJECT_INDEX: 'INDEX',
  OBJECT_LANGUAGE: 'LANGUAGE',
  OBJECT_LARGEOBJECT: 'LARGE OBJECT',
  OBJECT_MATVIEW: 'MATERIALIZED VIEW',
  OBJECT_OPCLASS: 'OPERATOR CLASS',
  OBJECT_OPERATOR: 'OPERATOR',
  OBJECT_OPFAMILY: 'OPERATOR FAMILY',
  OBJECT_POLICY: 'POLICY',
  OBJECT_PROCEDURE: 'PROCEDURE',
  OBJECT_PUBLICATION: 'PUBLICATION',
  OBJECT_ROLE: 'ROLE',
  OBJECT_ROUTINE: 'ROUTINE',
  OBJECT_RULE: 'RULE',
  OBJECT_SCHEMA: 'SCHEMA',
  OBJECT_SEQUENCE: 'SEQUENCE',
  OBJECT_STATISTIC_EXT: 'STATISTICS',
  OBJECT_SUBSCRIPTION: 'SUBSCRIPTION',
  OBJECT_TABCONSTRAINT: 'CONSTRAINT',
  OBJECT_TABLE: 'TABLE',
  OBJECT_TABLESPACE: 'TABLESPACE',
  OBJECT_TRANSFORM: 'TRANSFORM',
  OBJECT_TRIGGER: 'TRIGGER',
  OBJECT_TSCONFIGURATION: 'TEXT SEARCH CONFIGURATION',
  OBJECT_TSDICTIONARY: 'TEXT SEARCH DICTIONARY',
  OBJECT_TSPARSER: 'TEXT SEARCH PARSER',
  OBJECT_TSTEMPLATE: 'TEXT SEARCH TEMPLATE',
  OBJECT_TYPE: 'TYPE',
  OBJECT_VIEW: 'VIEW'
};

export const objtypeName = (arg) => {
  return TYPE_NAMES[arg];
};

export const getConstraintFromConstrType = (type) => {
  switch (type) {
    case 'CONSTR_NULL':
      return 'NULL';
    case 'CONSTR_NOTNULL':
      return 'NOT NULL';
    case 'CONSTR_DEFAULT':
      return 'DEFAULT';
    case 'CONSTR_CHECK':
      return 'CHECK';
    case 'CONSTR_PRIMARY':
      return 'PRIMARY KEY';
    case 'CONSTR_UNIQUE':
      return 'UNIQUE';
    case 'CONSTR_EXCLUSION':
      return 'EXCLUDE';
    case 'CONSTR_FOREIGN':
      return 'REFERENCES';

    case 'CONSTR_GENERATED':
      return 'GENERATED';

    case 'CONSTR_IDENTITY':
      return 'IDENTITY';

    case 'CONSTR_ATTR_DEFERRABLE':
    case 'CONSTR_ATTR_NOT_DEFERRABLE':
    case 'CONSTR_ATTR_DEFERRED':
    case 'CONSTR_ATTR_IMMEDIATE':
    default:
      return fail(type, 'ConstrType');
  }
};
