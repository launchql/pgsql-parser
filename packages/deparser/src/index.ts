import { Deparser } from "./deparser";

const deparse = Deparser.deparse;
export { deparse, Deparser };
