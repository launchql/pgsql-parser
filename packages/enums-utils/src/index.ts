import toInt from './enums2int.json';
import toStr from './enums2str.json';
import nodes from './nodes.json';

export { toInt, toStr, nodes };
export const getEnum = (enumType, key) => {
  if (typeof key === 'number') {
    return toStr[enumType][key];
  } else {
    return toInt[enumType][key];
  }
};

export * from './types';
