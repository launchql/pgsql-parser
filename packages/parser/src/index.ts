import { Deparser, deparse } from 'pgsql-deparser';
import {
  parseQuery,
  parseQuerySync,
  parsePlPgSQLSync as parseFunction
} from 'libpg-query';

function mapStmt({ stmt, stmt_len, stmt_location }) {
  return {
    RawStmt: {
      stmt,
      stmt_len,
      stmt_location: stmt_location || 0
    }
  };
}

export const parse = (sql) => {
  if (!sql) throw new Error('no SQL provided to parser');
  const result = parseQuerySync(sql);
  return result.stmts.map(mapStmt);
};

export const parseAsync = async (sql) => {
  if (!sql) throw new Error('no SQL provided to parser');
  const result = await parseQuery(sql);
  return result.stmts.map(mapStmt);
};

export { deparse, Deparser, parseFunction };
