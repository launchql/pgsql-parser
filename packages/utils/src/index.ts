import ast from './asts';
import nodes from './wrapped';
export { nodes };
export { ast };
export default {
    nodes,
    ast
}