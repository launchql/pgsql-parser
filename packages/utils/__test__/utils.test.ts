import * as t from '../src';
import { RangeVar, SelectStmt } from '@pgsql/types';
import { deparseSync as deparse } from 'pgsql-deparser';

it('simple SelectStmt', () => {
  const stmt: { SelectStmt: SelectStmt } = t.nodes.selectStmt({
    targetList: [
      t.nodes.resTarget({
        val: t.nodes.columnRef({
          fields: [t.nodes.aStar()]
        })
      })
    ],
    fromClause: [
      t.nodes.rangeVar({
        relname: 'some_table',
        inh: true,
        relpersistence: 'p'
      })
    ],
    limitOption: 'LIMIT_OPTION_DEFAULT',
    op: 'SETOP_NONE'
  });
  
  (stmt.SelectStmt.fromClause[0] as {RangeVar: RangeVar}).RangeVar.relname = 'another_table';

  expect(stmt).toMatchSnapshot();
  expect(deparse(stmt)).toMatchSnapshot();
});

it('SelectStmt with WHERE clause', () => {
  const selectStmt: { SelectStmt: SelectStmt } = t.nodes.selectStmt({
    targetList: [
      t.nodes.resTarget({
        val: t.nodes.columnRef({
          fields: [t.nodes.aStar()]
        })
      })
    ],
    fromClause: [
      t.nodes.rangeVar({
        schemaname: 'myschema',
        relname: 'mytable',
        inh: true,
        relpersistence: 'p'
      })
    ],
    whereClause: t.nodes.aExpr({
      kind: 'AEXPR_OP',
      name: [t.nodes.string({ sval: '=' })],
      lexpr: t.nodes.columnRef({
        fields: [t.nodes.string({ sval: 'a' })]
      }),
      rexpr: t.nodes.typeCast({
        arg: t.nodes.aConst({
          sval: t.ast.string({ sval: 't' })
        }),
        typeName: t.ast.typeName({
          names: [
            t.nodes.string({ sval: 'pg_catalog' }),
            t.nodes.string({ sval: 'bool' })
          ],
          typemod: -1
        })
      })
    }),
    limitOption: 'LIMIT_OPTION_DEFAULT',
    op: 'SETOP_NONE'
  });
  
  expect(deparse(selectStmt, {})).toEqual(`SELECT * FROM myschema.mytable WHERE a = CAST('t' AS boolean)`);
});

it('dynamic creation of tables', () => {
  // Example JSON schema
  const schema = {
    "tableName": "users",
    "columns": [
      { "name": "id", "type": "int", "constraints": ["PRIMARY KEY"] },
      { "name": "username", "type": "text" },
      { "name": "email", "type": "text", "constraints": ["UNIQUE"] },
      { "name": "created_at", "type": "timestamp", "constraints": ["NOT NULL"] }
    ]
  };

  // Construct the CREATE TABLE statement
  const createStmt = t.nodes.createStmt({
    relation: t.ast.rangeVar({ 
      relname: schema.tableName,
      inh: true,
      relpersistence: 'p'
    }),
    tableElts: schema.columns.map(column => t.nodes.columnDef({
      colname: column.name,
      typeName: t.ast.typeName({
        names: [t.nodes.string({ sval: column.type })]
      }),
      constraints: column.constraints?.map(constraint =>
        t.nodes.constraint({
          contype: constraint === "PRIMARY KEY" ? "CONSTR_PRIMARY" : constraint === "UNIQUE" ? "CONSTR_UNIQUE" : "CONSTR_NOTNULL"
        })
      )
    }))
  });

  // `deparse` function converts AST to SQL string
  const sql = deparse(createStmt, {});
  expect(sql).toMatchSnapshot();
})