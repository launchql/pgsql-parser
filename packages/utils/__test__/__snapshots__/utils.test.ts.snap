// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dynamic creation of tables 1`] = `"CREATE TABLE users (id int PRIMARY KEY, username text, email text UNIQUE, created_at timestamp NOT NULL)"`;

exports[`simple SelectStmt 1`] = `
{
  "SelectStmt": {
    "fromClause": [
      {
        "RangeVar": {
          "inh": true,
          "relname": "another_table",
          "relpersistence": "p",
        },
      },
    ],
    "limitOption": "LIMIT_OPTION_DEFAULT",
    "op": "SETOP_NONE",
    "targetList": [
      {
        "ResTarget": {
          "val": {
            "ColumnRef": {
              "fields": [
                {
                  "A_Star": {},
                },
              ],
            },
          },
        },
      },
    ],
  },
}
`;

exports[`simple SelectStmt 2`] = `"SELECT * FROM another_table"`;
